lib/system/web/src/keyboard_engine.rs

```diff
@@ -0,0 +1,170 @@
+use std::collections::{HashMap, BTreeSet};
+use std::cell::RefCell;
+use std::rc::Rc;
+use wasm_bindgen::JsCast;
+use wasm_bindgen::prelude::*;
+use itertools::Itertools;
+
+enum Action {
+    KeyUp,
+    KeyDown
+}
+
+type Bindings = HashMap<String, Box<dyn FnMut()>>;
+
+#[wasm_bindgen]
+pub struct KeyboardEngine {
+    bindings:        Rc<RefCell<Bindings>>,
+    _element:         web_sys::Element,
+    _prevent_default: bool,
+    //keydown_cb:    Closure<dyn FnMut(web_sys::KeyboardEvent)>,
+    //keyup_cb:      Closure<dyn FnMut(web_sys::KeyboardEvent)>,
+    _state:          Rc<RefCell<BTreeSet<String>>>
+}
+
+impl KeyboardEngine {
+    pub fn new(element: &web_sys::Element, prevent_default: bool) -> Self {
+        let state = Rc::new(RefCell::new(BTreeSet::new()));
+        let bindings = Rc::new(RefCell::new(HashMap::new()));
+        //let (keydown_cb, keyup_cb) = KeyboardEngine::init(element, bindings.clone(), state.clone());
+        KeyboardEngine::init(element, bindings.clone(), state.clone(), prevent_default);
```

Comment:
`bindings.clone()` this is misleading as it doesnt show the true cost of the code. Its better to write `Rc::clone(bindings)` instead, or define a method which clearly tells that this is cheap like `bindings.clone_ref()`. We usually do the latter, as it is more general (the former works only with `Rc`)